### PR TITLE
Add AssertApproximateSmallDataFrameEquality

### LINKED ISSUE #123: Row Comparer support Tolerance on nested Row
Row comparer currently only support tolerance on first layer of row. After that it will default to Row built in `equals` implementation

### LINKED ISSUE #124: [Performance Improvement] default to row equal implementation if tolerance is 0
using default `equals` implementation from Row should be much more performance. We can improve Dataframe equality assertion speed if strict comparison is required.

### PR DESCRIPTION
**Pull request by zeotuan**
- Add `assertApproximateSmallDataFrameEquality`
- Support approximate comparison within nested df
Close #123
Close #124

### CODE REVIEW

**Comment by MrPowers** [comment]
@zeotuan - thanks for helping on this project.  Feel free to ping me on LinkedIn (search "Matthew Powers CFA") or email me if you'd like to join the Slack where we chat!

**Comment by zeotuan** [comment]
Hi Matthew,

Thanks for creating such a useful tool! I'd love to join the Slack channel
and contribute further to improving this project and the Spark ecosystem.

Regards,
Tuan

**Comment by zeotuan** [comment]
Hi @SemyonSinchenko @MrPowers please help review. I can add some benchmark result if needed

**Review by SemyonSinchenko — COMMENTED** [review]
Thank you for the contribution! I left some comments

File: src/main/scala/com/github/mrpowers/spark/fast/tests/RowComparer.scala [hunk 1]
```diff
@@ -1,75 +1,44 @@
...
+          case f1: Float if o2.isInstanceOf[Float]   => Precision.equalsIncludingNaN(f1, o2.asInstanceOf[Float], tol)
+          case d1: Double if o2.isInstanceOf[Double] => Precision.equalsIncludingNaN(d1, o2.asInstanceOf[Double], tol)
+          case bd1: java.math.BigDecimal if o2.isInstanceOf[java.math.BigDecimal] =>
+            bd1.subtract(o2.asInstanceOf[java.math.BigDecimal]).abs().compareTo(new java.math.BigDecimal(tol)) == -1
+          case f1: Number if o2.isInstanceOf[Number] =>
+            val bd1 = new java.math.BigDecimal(f1.toString)
+            val bd2 = new java.math.BigDecimal(o2.toString)
+            bd1.subtract(bd2).abs().compareTo(new java.math.BigDecimal(tol)) == -1
+          case rr1: Row if o2.isInstanceOf[Row] => areRowsEqual(rr1, o2.asInstanceOf[Row], tol)
+          case _                                => o1 == o2
```
[the remark this answers is not in the document]
Comment (zeotuan) [reply src/main/scala/com/github/mrpowers/spark/fast/tests/RowComparer.scala:40]: added back Timestamp case. I have also added an Instant case for when spark.sql.datetime.java8API.enabled is true

**Review by SemyonSinchenko — APPROVED** [review]
LGTM! Thank you!

**Review by alfonsorr — APPROVED** [review]

**Pull request merged**